File: clients_sdk/python/bud_foundry/pipelines/tts.py
"""
BudTTS - Text-to-Speech pipeline
"""

from typing import Any, AsyncIterator, Callable, Optional

from ..types import TTSConfig, AudioEvent
from ..ws.session import WebSocketSession, SessionMetrics, ReconnectConfig
from ..rest.client import RestClient


class BudTTS:
    """Text-to-Speech pipeline for speech synthesis."""

    def __init__(
        self,
        url: str,
        api_key: Optional[str] = None,
        rest_client: Optional[RestClient] = None,
    ):
        """
        Initialize TTS pipeline.

        Args:
            url: WebSocket URL of the gateway (for streaming)
            api_key: Optional API key for authentication
            rest_client: Optional REST client for one-shot synthesis
        """
        self.url = url
        self.api_key = api_key
        self._rest_client = rest_client

    def create(
        self,
        config: Optional[TTSConfig] = None,
        provider: str = "deepgram",
        voice: Optional[str] = None,
        voice_id: Optional[str] = None,
        model: Optional[str] = None,
        sample_rate: int = 24000,
        reconnect: Optional[ReconnectConfig] = None,
    ) -> "TTSSession":
        """
        Create a TTS session.

        Args:
            config: Full TTS configuration (overrides other params)
            provider: TTS provider
            voice: Voice name
            voice_id: Voice ID
            model: Model to use
            sample_rate: Output sample rate
            reconnect: Reconnection configuration

        Returns:
            TTS session
        """
        if config is None:
            config = TTSConfig(
                provider=provider,
                voice=voice,
                voice_id=voice_id,
                model=model,
                sample_rate=sample_rate,
            )

        return TTSSession(
            url=self.url,
            api_key=self.api_key,
            config=config,
            reconnect=reconnect,
        )

    async def connect(
        self,
        config: Optional[TTSConfig] = None,
        provider: str = "deepgram",
        voice: Optional[str] = None,
        voice_id: Optional[str] = None,
        model: Optional[str] = None,
        sample_rate: int = 24000,
        reconnect: Optional[ReconnectConfig] = None,
    ) -> "TTSSession":
        """
        Create and connect a TTS session.

        Args:
            config: Full TTS configuration
            provider: TTS provider
            voice: Voice name
            voice_id: Voice ID
            model: Model to use
            sample_rate: Output sample rate
            reconnect: Reconnection configuration

        Returns:
            Connected TTS session
        """
        session = self.create(
            config=config,
            provider=provider,
            voice=voice,
            voice_id=voice_id,
            model=model,
            sample_rate=sample_rate,
            reconnect=reconnect,
        )
        await session.connect()
        return session

    async def synthesize(
        self,
        text: str,
        provider: str = "deepgram",
        voice: Optional[str] = None,
        voice_id: Optional[str] = None,
        model: Optional[str] = None,
        sample_rate: int = 24000,
    ) -> bytes:
        """
        Synthesize text to speech (one-shot, non-streaming).

        Args:
            text: Text to synthesize
            provider: TTS provider
            voice: Voice name
            voice_id: Voice ID
            model: Model to use
            sample_rate: Output sample rate

        Returns:
            Audio data as bytes (PCM)
        """
        if self._rest_client is None:
            # Derive REST URL from WebSocket URL
            rest_url = self.url.replace("ws://", "http://").replace("wss://", "https://")
            if rest_url.endswith("/ws"):
                rest_url = rest_url[:-3]
            self._rest_client = RestClient(base_url=rest_url, api_key=self.api_key)

        return await self._rest_client.speak(
            text=text,
            provider=provider,
            voice=voice,
            voice_id=voice_id,
            model=model,
            sample_rate=sample_rate,
        )


class TTSSession:
    """Active TTS session for streaming speech synthesis."""

    def __init__(
        self,
        url: str,
        api_key: Optional[str] = None,
        config: Optional[TTSConfig] = None,
        reconnect: Optional[ReconnectConfig] = None,
    ):
        """
        Initialize TTS session.

        Args:
            url: WebSocket URL
            api_key: API key
            config: TTS configuration
            reconnect: Reconnection config
        """
        self.config = config or TTSConfig(provider="deepgram")

        self._session = WebSocketSession(
            url=url,
            api_key=api_key,
            tts_config=self.config,
            reconnect=reconnect,
        )

        self._audio_handlers: list[Callable[[AudioEvent], Any]] = []
        self._playback_handlers: list[Callable[[], Any]] = []
        self._session.on("audio", self._on_audio)
        self._session.on("playback_complete", self._on_playback_complete)

    def _on_audio(self, event: AudioEvent) -> None:
        """Handle audio events."""
        for handler in self._audio_handlers:
            try:
                handler(event)
            except Exception:
                pass

    def _on_playback_complete(self, timestamp: Any) -> None:
        """Handle playback complete events."""
        for handler in self._playback_handlers:
            try:
                handler()
            except Exception:
                pass

    @property
    def connected(self) -> bool:
        """Whether the session is connected."""
        return self._session.connected

    @property
    def stream_id(self) -> Optional[str]:
        """Get the stream ID."""
        return self._session.stream_id

    def on(self, event: str, handler: Callable[..., Any]) -> None:
        """
        Register an event handler.

        Args:
            event: Event name (audio, playback_complete, error, close)
            handler: Event handler function
        """
        if event == "audio":
            self._audio_handlers.append(handler)
        elif event == "playback_complete":
            self._playback_handlers.append(handler)
        else:
            self._session.on(event, handler)

    def off(self, event: str, handler: Optional[Callable[..., Any]] = None) -> None:
        """Remove an event handler."""
        if event == "audio":
            if handler is None:
                self._audio_handlers.clear()
            elif handler in self._audio_handlers:
                self._audio_handlers.remove(handler)
        elif event == "playback_complete":
            if handler is None:
                self._playback_handlers.clear()
            elif handler in self._playback_handlers:
                self._playback_handlers.remove(handler)
        else:
            self._session.off(event, handler)

    async def connect(self, timeout: float = 10.0) -> None:
        """Connect to the gateway."""
        await self._session.connect(timeout=timeout)

    async def disconnect(self) -> None:
        """Disconnect from the gateway."""
        await self._session.disconnect()

    async def speak(
        self,
        text: str,
        flush: bool = False,
        allow_interruption: bool = True,
    ) -> None:
        """
        Synthesize and stream text to speech.

        Args:
            text: Text to synthesize
            flush: Whether to flush the TTS buffer immediately
            allow_interruption: Whether this TTS can be interrupted
        """
        await self._session.speak(text, flush=flush, allow_interruption=allow_interruption)

    async def clear(self) -> None:
        """Clear/stop current TTS playback."""
        await self._session.clear()

    def get_metrics(self) -> SessionMetrics:
        """Get session metrics."""
        return self._session.get_metrics()

    def reset_metrics(self) -> None:
        """Reset session metrics."""
        self._session.reset_metrics()

    async def __aiter__(self) -> AsyncIterator[AudioEvent]:
        """Iterate over audio events."""
        async for message in self._session:
            if message.get("type") == "audio":
                yield message["audio"]

    async def __aenter__(self) -> "TTSSession":
        """Async context manager entry."""
        await self.connect()
        return self

    async def __aexit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
        """Async context manager exit."""
        await self.disconnect()

File: clients_sdk/typescript/src/pipelines/tts.ts
/**
 * BudTTS Pipeline
 * Text-to-Speech pipeline for real-time synthesis
 */

import type { TTSConfig } from '../types/config.js';
import type { FeatureFlags } from '../types/features.js';
import type { AudioEvent } from '../ws/events.js';
import { BasePipeline, type BasePipelineConfig } from './base.js';
import { PCMPlayer, type PlayerConfig } from '../audio/player.js';
import { RestClient } from '../rest/client.js';

/**
 * BudTTS configuration
 */
export interface BudTTSConfig extends BasePipelineConfig {
  /** TTS provider (e.g., 'elevenlabs', 'deepgram', 'azure') */
  provider?: string;
  /** Voice name or identifier */
  voice?: string;
  /** Voice ID (provider-specific) */
  voiceId?: string;
  /** Model to use (e.g., 'eleven_turbo_v2') */
  model?: string;
  /** Output sample rate (default: 24000) */
  sampleRate?: number;
  /** Output audio format (default: 'linear16') */
  audioFormat?: string;
  /** Speech rate multiplier (default: 1.0) */
  speed?: number;
  /** Pitch adjustment (default: 1.0) */
  pitch?: number;
  /** Volume adjustment (default: 1.0) */
  volume?: number;
  /** Voice stability (ElevenLabs specific, 0-1) */
  stability?: number;
  /** Voice similarity boost (ElevenLabs specific, 0-1) */
  similarityBoost?: number;
  /** Voice style (ElevenLabs specific, 0-1) */
  style?: number;
  /** Use speaker boost (ElevenLabs specific) */
  useSpeakerBoost?: boolean;
  /** Whether to auto-play received audio (default: true) */
  autoPlay?: boolean;
  /** Player configuration */
  playerConfig?: PlayerConfig;
  /** Base URL for REST API (used for one-shot synthesis) */
  restBaseUrl?: string;
}

/**
 * BudTTS - Text-to-Speech Pipeline
 */
export class BudTTS extends BasePipeline {
  private ttsConfig: TTSConfig;
  private player: PCMPlayer | null = null;
  private autoPlay: boolean;
  private restClient: RestClient | null = null;
  private isSpeaking = false;
  private speakQueue: string[] = [];
  private isProcessingQueue = false;

  constructor(config: BudTTSConfig) {
    const ttsConfig: TTSConfig = {
      provider: config.provider ?? 'deepgram',
      voice: config.voice,
      voiceId: config.voiceId,
      model: config.model,
      sampleRate: config.sampleRate ?? 24000,
      audioFormat: config.audioFormat ?? 'linear16',
      speed: config.speed,
      pitch: config.pitch,
      volume: config.volume,
      stability: config.stability,
      similarityBoost: config.similarityBoost,
      style: config.style,
      useSpeakerBoost: config.useSpeakerBoost,
    };

    super({
      ...config,
      sessionConfig: {
        tts: ttsConfig,
      },
    });

    this.ttsConfig = ttsConfig;
    this.autoPlay = config.autoPlay ?? true;

    // Setup audio player if auto-play enabled
    if (this.autoPlay) {
      this.player = new PCMPlayer({
        sampleRate: ttsConfig.sampleRate,
        ...config.playerConfig,
      });
    }

    // Setup REST client for one-shot synthesis
    if (config.restBaseUrl) {
      this.restClient = new RestClient({
        baseUrl: config.restBaseUrl,
        apiKey: config.apiKey,
      });
    }

    // Forward audio events
    this.session.on('audio', (e) => this.handleAudio(e));
    this.session.on('speaking', (e) => {
      this.isSpeaking = e.speaking;
      this.emitter.emit('speaking', e);
    });
  }

  /**
   * Handle received audio
   */
  private handleAudio(event: AudioEvent): void {
    this.emitter.emit('audio', event);

    if (this.autoPlay && this.player) {
      this.player.addPCM(event.audio);
    }
  }

  /**
   * Initialize the player
   */
  async initializePlayer(): Promise<void> {
    if (this.player) {
      await this.player.initialize();
    }
  }

  /**
   * Speak text
   * @param text Text to synthesize
   * @param options Optional speak options
   */
  async speak(text: string, options?: {
    voice?: string;
    voiceId?: string;
    provider?: string;
    model?: string;
    speed?: number;
    pitch?: number;
    flush?: boolean;
  }): Promise<void> {
    if (!this.isConnected()) {
      await this.connect();
    }

    // Ensure player is initialized
    if (this.autoPlay && this.player) {
      await this.player.initialize();
    }

    this.session.speak(text, options);
  }

  /**
   * Queue text for speaking
   * @param text Text to queue
   */
  queueSpeak(text: string): void {
    this.speakQueue.push(text);
    this.processQueue();
  }

  /**
   * Process speak queue
   */
  private async processQueue(): Promise<void> {
    if (this.isProcessingQueue || this.isSpeaking || this.speakQueue.length === 0) {
      return;
    }

    this.isProcessingQueue = true;

    while (this.speakQueue.length > 0 && !this.isSpeaking) {
      const text = this.speakQueue.shift();
      if (text) {
        await this.speak(text);
        // Wait for speaking to finish
        await this.waitForSpeakingDone();
      }
    }

    this.isProcessingQueue = false;
  }

  /**
   * Wait for current speaking to finish
   */
  private waitForSpeakingDone(): Promise<void> {
    return new Promise((resolve) => {
      if (!this.isSpeaking) {
        resolve();
        return;
      }

      const handler = (e: { speaking: boolean }) => {
        if (!e.speaking) {
          this.off('speaking', handler);
          resolve();
        }
      };

      this.on('speaking', handler);
    });
  }

  /**
   * One-shot synthesis using REST API
   * @param text Text to synthesize
   * @param options Optional options
   * @returns Audio buffer
   */
  async synthesize(text: string, options?: {
    provider?: string;
    voice?: string;
    voiceId?: string;
    model?: string;
    sampleRate?: number;
    format?: string;
  }): Promise<ArrayBuffer> {
    if (!this.restClient) {
      throw new Error('REST client not configured. Provide restBaseUrl in config.');
    }

    const result = await this.restClient.speak(text, {
      provider: options?.provider ?? this.ttsConfig.provider,
      voice: options?.voice ?? this.ttsConfig.voice,
      model: options?.model ?? this.ttsConfig.model,
      sampleRate: options?.sampleRate ?? this.ttsConfig.sampleRate,
      format: options?.format ?? this.ttsConfig.audioFormat,
    });

    return result.audio;
  }

  /**
   * Update TTS configuration
   */
  updateConfig(config: Partial<TTSConfig>): void {
    this.ttsConfig = { ...this.ttsConfig, ...config };
    this.session.updateTTSConfig(config);
  }

  /**
   * Stop current speech
   */
  stopSpeaking(): void {
    this.session.interrupt();
    this.player?.stop();
    this.speakQueue = [];
    this.isSpeaking = false;
  }

  /**
   * Pause playback
   */
  pausePlayback(): void {
    this.player?.pause();
  }

  /**
   * Resume playback
   */
  resumePlayback(): void {
    this.player?.play();
  }

  /**
   * Clear playback buffer
   */
  clearBuffer(): void {
    this.player?.clearBuffer();
  }

  /**
   * Set playback volume
   * @param gainDb Volume in decibels
   */
  setVolume(gainDb: number): void {
    this.player?.setGain(gainDb);
  }

  /**
   * Check if currently speaking
   */
  getIsSpeaking(): boolean {
    return this.isSpeaking;
  }

  /**
   * Get current queue length
   */
  getQueueLength(): number {
    return this.speakQueue.length;
  }

  /**
   * Dispose resources
   */
  async dispose(): Promise<void> {
    this.stopSpeaking();
    if (this.player) {
      await this.player.dispose();
      this.player = null;
    }
    await this.disconnect();
  }
}

/**
 * Create a BudTTS instance
 */
export function createBudTTS(config: BudTTSConfig): BudTTS {
  return new BudTTS(config);
}
